File: iprep/dynamic_programming/max_array_sum.py
#!/bin/python3

import math
import os
import random
import re
import sys

# Complete the maxSubsetSum function below.
def maxSubsetSum(arr):

    S = [0 for a in arr]

    for i, a in enumerate(arr):
        choices = []
        for j in range(0, i - 1):
            choices.append(S[j]+arr[i])
        for j in range(0, i-1):
            choices.append(arr[j]+arr[i])
        if len(choices) > 0:
            S[i] = max(choices)

    return max(S)

File: iprep/dynamic_programming/max_array_sum.scala
import java.io._
import java.math._
import java.security._
import java.text._
import java.util._
import java.util.concurrent._
import java.util.function._
import java.util.regex._
import java.util.stream._

object Solution {

    // Complete the maxSubsetSum function below.
    def maxSubsetSum(arr: Array[Int]): Int = {

        if (arr.size == 1) {
            return arr.last
        }

        if (arr.size == 2) {
            return Math.max(arr.sum, arr.max)
        }
        val D = Array.fill(arr.size+2)(0)

        for (i <-2 until D.size) {

            D(i) = Math.max(D(i-2)+arr(i-2), D(i-1))
        }

        return D.last
    }

    def main(args: Array[String]) {
        val stdin = scala.io.StdIn

        val printWriter = new PrintWriter(sys.env("OUTPUT_PATH"))

        val n = stdin.readLine.trim.toInt

        val arr = stdin.readLine.split(" ").map(_.trim.toInt)
        val res = maxSubsetSum(arr)

        printWriter.println(res)

        printWriter.close()
    }
}
